keps/sig-node/4216-image-pull-per-runtime-class/README.md

```diff
@@ -0,0 +1,1338 @@
+         "mediaType": "application/vnd.docker.distribution.manifest.v2+json",
+         "size": 2007,
+         "digest": "sha256:f265d2f398ffce7252d6162ead0bc802afad2de309cf662ec16645d1e0e85564",
+         "platform": {
+            "architecture": "s390x",
+            "os": "linux"
+         }
+      },
+      {
+         "mediaType": "application/vnd.docker.distribution.manifest.v2+json",
+         "size": 2831,
+         "digest": "sha256:53c5f0dd905eef3899284d845431ccaa1045f97fc205edd87dfc2151c4331980",
+         "platform": {
+            "architecture": "amd64",
+            "os": "windows",
+           <mark> "os.version": "10.0.20348.1970"  </mark>
+         }
+      },
+      {
+         "mediaType": "application/vnd.docker.distribution.manifest.v2+json",
+         "size": 2831,
+         "digest": "sha256:5981df14a07aaa7fe0c7d80a4c61f33f4ad4d8d29a346fd1b2cacf090b3de8c2",
+         "platform": {
+            "architecture": "amd64",
+            "os": "windows",
+           <mark> "os.version": "10.0.17763.4851" </mark>
+         }
+      }
+   ]
+}
+</pre>
+
+Containerd's client structure has pointer to the default platform matcher based on what OS/platform the node is on (Linux,windows, Darwin etc). The platform matcher on linux attempts to check for host and guest OS match and optionally for a variant match as well. However, the platform matcher on windows checks for exact OS version match between the host and guest (OSVersion field in the oci ImageSpec is primarily used only for windows manifests) and if a manifest with such an exact match is found, it is pulled. Otherwise an error is thrown. This is the desired behavior for process isolated windows containers as its shares the kernel with the host OS. However, it is not ideal for hyperV isolated windows containers. This is because OS version of the host and the UVM image need not be the same. For example, we could have a UVM running Windows Server 2019 image on a Windows Server 2022 host and this should be allowed. For the full matrix of accepted behavior, check the link here: https://learn.microsoft.com/en-us/virtualization/windowscontainers/deploy-containers/version-compatibility?tabs=windows-server-2022%2Cwindows-10#windows-server-host-os-compatibility 
+
+This problem can be fixed if containerd, CRI and kubelet refer to images as a tuple of (imageName/ID, runtimeClass) instead of the imageName/ID. With this, we can specify the runtimeClass a particular image is being run for and pull the appropriate manifest from the image index. This document details the changes needed to support pulling images per runtime class.
+Adding this support would enable us to have different versions of the same image on a given node and could be really beneficial in virtualization based runtimes. In windows, differnt versions of a multiarch image could be pulled based on hyperV and process isolation. It will also be very helpful to support future projects to support snapshotter per runtimeclass. 
```

Comment:
```suggestion
Adding this support would enable us to have different versions of the same image on a given node and could be really beneficial in virtualization based runtimes. In windows, differnt versions of a multiarch image could be pulled based on hyperV and process isolation. It will also be very helpful to support future projects to support snapshotter per runtimeclass. 
```

It would be helpful to give a few reasons why this is helpful to add additional context.

Reply:
The motivation section has more details with examples.